### ISSUE TITLE
Allowing scopes to access request.auth?

### ISSUE THREAD

**Issue by jamesdixon** [hapijs/hapi#2831]
Hi,

I started using dynamic scopes and they were working really great when we were passing an `id` in via the url. However, we are now taking a different approach where we're no longer passing an `id` in via the url and now just pulling the user's id from their credentials which are available in `request.auth.credentials`. I'd prefer not to add authorization checks to my handlers as I find scopes to be much cleaner and easier to read. All that said, I wanted to propose allowing scopes to access `request.auth.credentials` in addition to `request.payload` and `request.query`. It's possible I may be missing something, but I didn't see any security issues with doing so offhand.

Thoughts? 

Also, if there is a better way than what I'm suggesting, please feel free to educate me :+1: 

Thanks!
James

**Comment by hueniverse** [comment]
Payload is not available at that point in the lifecycle. Query makes little sense - can you provide a use case where it makes sense to link the scope with a query value?

As for the credentials, that is a weird design. If the scope is in the credentials, it should be under the  `scope` key and should not be dynamic on the route.

Allowing path parameters in dynamic scope makes sense because it allows scoping of resources. But query, payload, and credentials does not make sense to me.

**Comment by jamesdixon** [comment]
> Payload is not available at that point in the lifecycle.

You're absolutely correct. That was a typo on my part; I was confusing `payload` with `params`.

> Query makes little sense - can you provide a use case where it makes sense to link the scope with a query value?

I just realized my ask came off wrong. I was asking to allow access to `request.auth`, but not `request.payload` or `request.query`. My language suggested that I wanted all three, but in reality I was thinking that access to `query` and `payload` (`params`) already existed. Sorry about that. Although, unless I'm misunderstanding,  according to the docs it seems that scopes do have access to `query`. See [Route options > scopes](http://hapijs.com/api#route-options).

> As for the credentials, that is a weird design. If the scope is in the credentials, it should be under the scope key and should not be dynamic on the route.
>
> Allowing path parameters in dynamic scope makes sense because it allows scoping of resources. But query, payload, and credentials does not make sense to me.

It very well may be a weird design. Let me lay out an example to see if it makes more sense with what I'm trying to accomplish:

Once a user has been authenticated, we store some information such as `id` in `request.auth.credentials`. Previously, we also stored the dynamic scope. So for example, an owner of a business might have the `business-1` scope and we would scope a resource such as `businesses/{id}` using a dynamic scope. Doing that we could prevent an authenticated, but unauthorized user, from accessing another businesses' data.

Now we've moved away from always passing in the `id` of the resource in the URL and instead are grabbing it from the `request.auth.credentials` object since that information is already there. Unfortunately, because it's no longer passed in the URL, I can take advantage of scoping the resources using a dynamic scope.

My thought was that if I could do something like

```javascript
// route config
...
scope: [ 'business-{auth.credentials.businessId}' ]
...
```
which would still allow me to dynamically scope resources at the route level and avoid having to add some sort of other authorization mechanism elsewhere. I realize that the scope already exists in the credentials object, but not sure how I could use that to dynamically scope the resource.

Thanks, @hueniverse
